src/librustdoc/clean/mod.rs

```diff
@@ -2109,9 +2071,11 @@ pub(crate) fn clean_middle_ty<'tcx>(
             // FIXME: should we merge the outer and inner binders somehow?
```

Comment:
I'm not super familiar with some of the details of bound vars yet, continuously working towards that goal. I can't tell you all possible cases where the type and the signature have bound vars.

I can give you one example though which rustdoc renders just fine:

```rs
pub fn f(_: impl for<'a> Trait<for<'b> fn(&'a (), &'b ())>) {  }
pub trait Trait<T> {}
```

Reply:
Equipped with my limited understanding, I disagree with the comment, it would be super odd to merge different binder lists, rustc does it like one or a few times in exceptional cases. There can be semantic difference between `for<'a> … for<'b> …` and `for<'a, 'b>`. rustc needs to shift the de Bruijn index and the var index for correctness.

A few months ago, I chatted with compiler-errors a bit about `ty::Binder<_>` and bound vars in rustdoc and from what I've learned rustdoc (still) doesn't properly track bound vars. This was relevant for #112463 where we don't have enough information to be able to deal with (escaping) late bound vars. I haven't understood everything yet and thus my explanation sucks.

I'm pretty sure from what I've learned so far, proper bound var handling would prevent ICEs in rustdoc like #108503 and probably other ones like https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3AT-rustdoc%20label%3AI-ICE%20label%3AA-auto-traits%20label%3AA-synthetic-impls.